MON-1687 - Add new monitoring content for remote write configuration (#36054)
openshift/openshift-docs: modules/monitoring-configuring-remote-write.adoc

// Module included in the following assemblies:
//
// * monitoring/configuring-the-monitoring-stack.adoc

[id="configuring_remote_write_storage_{context}"]
= Configuring remote write storage

You can configure remote write to enable Prometheus to send ingested metrics to remote systems for storage. 

> Doing so has no impact on how or for how long Prometheus stores metrics using local storage.

Reviewer: 'local storage' as a term has special implications. Consider using the term `defined storage` or `configured storage medium`.

Author: Good point. Will update the wording to clarify,

Author: per QE, changing to `persistent storage`

.Prerequisites

> * *If you are configuring core OKD monitoring components:*

Reviewer: Don't use OKD.

Reviewer 2: Are there precedents we can follow? Searching the docs for `monitoring components`, I see instances of `*If you are configuring core {product-title} monitoring components*:`

Author: Thanks. Will update accordingly.

Author: Fixed

** You have access to the cluster as a user with the `cluster-admin` role.
** You have created the `cluster-monitoring-config` `ConfigMap` object.
* *If you are configuring components that monitor user-defined projects:*

> ** You have access to the cluster as a user with the `cluster-admin` role or as a user with the `user-workload-monitoring-config-edit` role in the `openshift-user-workload-monitoring` project.

Reviewer: Is this a duplicate 'requirement'? How can / could this be better framed/reframed to simplify the instructions?

Reviewer 2: These lines are copied from and have precedents in various other files, such as `modules/monitoring-configuring-a-local-persistent-volume-claim.adoc`. It would be better to stick with precedent for this PR, and create a separate enhancement request to improve the wording and content structure across all the precedents.

Author: @sferich888, you have a good point, but I agree with @rolfedh here for this PR. I'm leaving the wording as-is for now to stay consistent with the rest of the docs.

** You have created the `user-workload-monitoring-config` `ConfigMap` object.

> * You have installed the OpenShift CLI (`oc`).

Reviewer: @rolfedh do we list this on other pre-requsites? Is this implied?

Reviewer 2: Yes, this line is used by 74 instances of other files throughout the documentation.

Author: I'm leaving the wording as-is to stay consistent with the rest of the docs.

> * You have set up a remote write compatible endpoint (such as Thanos) and know the endpoint URL.

Reviewer: What are valid 'remote write compatible endpoints'? Can we define or outline the requirements of what is acceptable here better?

Reviewer: What specification or RFC is our API providing?

Author: @jan--f - can you provide more detail to address these questions?

Reviewer 3: The actual specification is still somewhat being worked out. See https://docs.google.com/document/d/1LPhVRSFkGNSuU1fBd81ulhsCPR4hkSZyyBj1SZ8fWOM/edit#heading=h.3p42p5s8n0ui
Its probably not too useful to users to have a link to the protocol specification, at least not at this point.

What is the goal of putting this into the user facing docs? Maybe we can find an alternative solution, say linking to the l[ist of `remote_write` compatible endpoints](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage)?

Author: Agree with @jan--f . Added the suggested to link to the remote_write compatible endpoints.

* You have set up authentication credentials for the remote write endpoint. 
+
[CAUTION]
====
To reduce security risks, avoid sending metrics to an endpoint via unencrypted HTTP or without using authentication.
====

.Procedure

. Edit the `cluster-monitoring-config` `ConfigMap` object in the `openshift-monitoring` project:
+
[source,terminal]
----
$ oc -n openshift-monitoring edit configmap cluster-monitoring-config
----

. Add a `remoteWrite:` section under `data/config.yaml/prometheusK8s`.

> . In this section, add an endpoint URL, authetication credentials for the endpoint, and write label configurations.

Reviewer 3: ```suggestion
. In this section, add an endpoint URL, authentication credentials for the endpoint, and write relabel configurations.
```

Reviewer 2: ```suggestion
```
I think you can remove this line and promote step 3.a to step 3, and step 3.b to step 4.

Author: Updated. Thanks.

> .. Add an endpoint URL and authentication credentials:

Reviewer 2: ```suggestion
. Add an endpoint URL and authentication credentials:
```

Author: Fixed. Thanks

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote.write.endpoint"

>         <endpoint_authentication_credentials>

Reviewer: Can this be detailed out better?

Author: Added code samples for both types of supported auth methods.

----
+

> For `endpoint_authentication_credentials` substitute your credentials.

Reviewer: Should we enumerate these better to list the types of authentication mechanisms we support? We list Basic Authentication and TLS authentication in examples but are these the only ones?

Reviewer: Should Authentication be its own 'section'? cc: @rolfedh

Reviewer 2: Let's see how the proposed updates to the structure work. If they don't disrupt the flow too much, it might be better for users to deliver just-in-time information this way.

Author: @sferich888 - according to @jan--f, TLS Authentication and Basic Authentication are the only 2 protocols we are supporting. We should add a sample with values for tlsConfig. Jan, can you please provide a sample?

Author: Added samples for both auth methods.

Currently supported authentication methods are basic authentication (`basicAuth`) and client TLS (`tlsConfig`) authentication.
+
* The following example configures basic authentication:
+
[source,yaml]
----
        basicAuth:
          username:
            <usernameSecret>
          password:
            <passwordSecret>
----
Substitute `<usernameSecret>` and `<passwordSecret>` accordingly.
+
* The following example configures client TLS authentication:
+
[source,yaml]
----
         tlsConfig:
           cert:
              <certSecret>
           keySecret:
              <keySecret>
----
Substitute `<certSecret>` and `<keySecret>` accordingly.
+

> .. Add write relabel configuration values after the authentication credentials:

Reviewer: Should the 'write relabel configuration' be its own section? cc: @rolfedh

Reviewer 2: ```suggestion
. Add write relabel configuration values after the authentication credentials:
```

Author: I am restructuring the procedure so that the authentication information is separated more distinctly from the write relabel step.

Author: Updated.

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote.write.endpoint"
        <endpoint_authentication_credentials>
        <write_relabel_configs>
----
+
For `<write_relabel_configs>` substitute a list of write relabel configurations for metrics that you want to send to the remote endpoint.
+
See the link:https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config[Prometheus relabel_config documentation] for information about write relabel configuration options.
+
The following example `cluster-monitoring-config` `ConfigMap` settings show how to:
+
* forward a single metric called `my_metric`
+

> * configure basic authentication using the `remoteWriteAuth` secret with the `user` and `password` keys, which would contain the endpoint authentication credentials

Reviewer: Did we just mix auth and rewire label config examples? Could these be split for reader clarity?

Author: This issue should be cleared up after I restructure the procedure.

Author: Updated. Thanks.

  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote.write.endpoint"
        writeRelabelConfig:
        - source_labels: [__name__]
          regex: 'my_metric'
          action: keep
        basicAuth:
          username:
            name: remoteWriteAuth
            key: user
          password:
            name: remoteWriteAuth
            key: password

----

. If required, configure remote write for the Prometheus instance that monitors user-defined projects by changing the `name` and `namespace` `metadata` values as follows:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    prometheus:
      remoteWrite:
      - url: "https://remote.write.endpoint"
        <endpoint_authentication_credentials>
        <write_relabel_configs>
----
+
[NOTE]
====
The Prometheus config map component is called `prometheusK8s` in the `cluster-monitoring-config` `ConfigMap` object and `prometheus` in the `user-workload-monitoring-config` `ConfigMap` object.
====
+

> The other config map components in the `data/config.yaml/prometheus/remoteWrite` section remain unchanged.

Reviewer: What other components? Do we need to say this?

Author: Good point. I think this sentence can be omitted.

// Module included in the following assemblies:
//
// * monitoring/configuring-the-monitoring-stack.adoc

[id="configuring_remote_write_storage_{context}"]
= Configuring remote write storage

You can configure remote write to enable Prometheus to send ingested metrics to remote systems for long-term storage. 

> Doing so has no impact on how or for how long Prometheus stores metrics using persistent storage.

Author: Updated.

> Doing so has no impact on how or for how long Prometheus stores metrics using persistent local storage.

Reviewer 4: `persistent local storage`
change to
`persistent storage` is better

Author: Updated. Thanks!

* You have installed the OpenShift CLI (`oc`).
* You have set up a remote write compatible endpoint (such as Thanos) and know the endpoint URL.
See the link:https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage[Prometheus remote endpoints and storage documentation] for information about endpoints that are compatible with the remote write feature.
* You have set up authentication credentials for the remote write endpoint. 
+
[CAUTION]
====
To reduce security risks, avoid sending metrics to an endpoint via unencrypted HTTP or without using authentication.
====

.Procedure

. Edit the `cluster-monitoring-config` `ConfigMap` object in the `openshift-monitoring` project:
+
[source,terminal]
----
$ oc -n openshift-monitoring edit configmap cluster-monitoring-config
----

. Add a `remoteWrite:` section under `data/config.yaml/prometheusK8s`.

. Add an endpoint URL and authentication credentials in this section:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:

>       - url: "https://remote.write.endpoint"

Reviewer 4: `- url: "https://remote.write.endpoint"`
change to
`- url: "https://remote-write.endpoint"`
is better

----
+
For `endpoint_authentication_credentials` substitute the endpoint's credentials.
Currently supported authentication methods are basic authentication (`basicAuth`) and client TLS (`tlsConfig`) authentication.
+
* The following example configures basic authentication:
+
[source,yaml]
----
basicAuth:
  username:
    <usernameSecret>
  password:
    <passwordSecret>
----
Substitute `<usernameSecret>` and `<passwordSecret>` accordingly. The following sample shows basic authentication configured with `remoteWriteAuth` for the `name` values and `user` and `password` for the `key` values, which contain the endpoint authentication credentials:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote.write.endpoint"
        basicAuth:
          username:
            name: remoteWriteAuth
            key: user
          password:
            name: remoteWriteAuth
            key: password
----
+
* The following example configures client TLS authentication:
+
[source,yaml]
----
tlsConfig:
  cert:
    <certSecret>

>   keySecret:

Reviewer 4: should be key
see from: https://github.com/openshift/cluster-monitoring-operator/blob/release-4.9/pkg/manifests/config.go#L206

    <keySecret>
----
Substitute `<certSecret>` and `<keySecret>` accordingly. The following sample shows a TLS authentication configuration using `selfsigned-mtls-bundle` for the `name` values and `ca.crt` for the `ca` `key` value, `client.crt` for the `cert` `key` value, and `client.key` for the `keySecret` `key` value:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote.write.endpoint"
      tlsConfig:
        ca:
          secret:
            name: selfsigned-mtls-bundle
            key: ca.crt
        cert:
          secret:
            name: selfsigned-mtls-bundle
            key: client.crt

>         keySecret:

Reviewer 4: should be `key`
see from: https://github.com/openshift/cluster-monitoring-operator/blob/release-4.9/pkg/manifests/config.go#L206

Reviewer 3: Same as above.

Author: @jan--f and @juzhao : I'm not sure what's correct here. Should `keySecret` be changed to `key` or not?

Reviewer 4: @bburt-rh no need, `keySecret` is right

          name: selfsigned-mtls-bundle
          key: client.key
----

. Add write relabel configuration values after the authentication credentials:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote.write.endpoint"
        <endpoint_authentication_credentials>
        <write_relabel_configs>
----
+
For `<write_relabel_configs>` substitute a list of write relabel configurations for metrics that you want to send to the remote endpoint.
+
The following example shows how to forward a single metric called `my_metric`:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote.write.endpoint"

>         writeRelabelConfig:

Reviewer 4: I think it should be `writeRelabelConfigs`, see from:
https://github.com/openshift/cluster-monitoring-operator/pull/1308/files#diff-420c938cae65d6e9c4d4e386d768532276257149dbcaf476a90884a29fe36bd4R154

Reviewer 3: Good catch, this is my mistake in the document I provided @bburt-rh with. Fixed in the the gdoc as well.

Author: Updated in the doc.

        - source_labels: [__name__]
          regex: 'my_metric'
          action: keep

----
+
See the link:https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config[Prometheus relabel_config documentation] for information about write relabel configuration options.

. If required, configure remote write for the Prometheus instance that monitors user-defined projects by changing the `name` and `namespace` `metadata` values as follows:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    prometheus:
      remoteWrite:
      - url: "https://remote.write.endpoint"
        <endpoint_authentication_credentials>
        <write_relabel_configs>
----
+
[NOTE]
====
The Prometheus config map component is called `prometheusK8s` in the `cluster-monitoring-config` `ConfigMap` object and `prometheus` in the `user-workload-monitoring-config` `ConfigMap` object.
====

. Save the file to apply the changes to the `ConfigMap` object.
The pods affected by the new configuration restart automatically.
+
[NOTE]
====

> Configurations applied to the `user-workload-monitoring-config` `ConfigMap` object are not activated unless a cluster administrator has enabled monitoring for user-defined projects.

Reviewer 4: `a cluster administrator`
->
`the cluster administrator`

Reviewer 3: There can be multiple administrators no? `a cluster administrator` should be fine imho.

Author: I agree with @jan--f. Leaving as-is.

// Module included in the following assemblies:
//
// * monitoring/configuring-the-monitoring-stack.adoc

[id="configuring_remote_write_storage_{context}"]
= Configuring remote write storage

> You can configure remote write to enable Prometheus to send ingested metrics to remote systems for long-term storage. 

Author: Revised per suggestion

* You have set up a remote write compatible endpoint (such as Thanos) and know the endpoint URL.
See the link:https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage[Prometheus remote endpoints and storage documentation] for information about endpoints that are compatible with the remote write feature.
* You have set up authentication credentials for the remote write endpoint. 
+
[CAUTION]
====
To reduce security risks, avoid sending metrics to an endpoint via unencrypted HTTP or without using authentication.
====

.Procedure

. Edit the `cluster-monitoring-config` `ConfigMap` object in the `openshift-monitoring` project:
+
[source,terminal]
----
$ oc -n openshift-monitoring edit configmap cluster-monitoring-config
----

. Add a `remoteWrite:` section under `data/config.yaml/prometheusK8s`.

. Add an endpoint URL and authentication credentials in this section:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote-write.endpoint"
        <endpoint_authentication_credentials>
----
+

> For `endpoint_authentication_credentials` substitute the endpoint's credentials.

Author: Revised per suggestion

Currently supported authentication methods are basic authentication (`basicAuth`) and client TLS (`tlsConfig`) authentication.
+
* The following example configures basic authentication:
+
[source,yaml]
----
basicAuth:
  username:
    <usernameSecret>
  password:
    <passwordSecret>
----

> Substitute `<usernameSecret>` and `<passwordSecret>` accordingly. The following sample shows basic authentication configured with `remoteWriteAuth` for the `name` values and `user` and `password` for the `key` values, which contain the endpoint authentication credentials:

Reviewer 5: The previous sentence is about the preceding sample. Suggest starting the 'The following sample...' sentence on a new paragraph to indicate the switch in the context.
Also, suggest breaking the second sentence into two sentences.

Author: @Preeticp: Revised per your suggestion

            key: ca.crt
        cert:
          secret:
            name: selfsigned-mtls-bundle
            key: client.crt
        keySecret:
          name: selfsigned-mtls-bundle
          key: client.key
----

. Add write relabel configuration values after the authentication credentials:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote-write.endpoint"
        <endpoint_authentication_credentials>
        <write_relabel_configs>
----
+
For `<write_relabel_configs>` substitute a list of write relabel configurations for metrics that you want to send to the remote endpoint.
+
The following example shows how to forward a single metric called `my_metric`:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote-write.endpoint"
        writeRelabelConfigs:
        - source_labels: [__name__]
          regex: 'my_metric'
          action: keep

----
+

> See the link:https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config[Prometheus relabel_config documentation] for information about write relabel configuration options.

Reviewer 5: @sferich888 is this external link Ok to be included in the docs?
